### PR TITLE
Handle missing pytest_cov gracefully

### PR DESCRIPTION
**Pull request by cottsay**
Intent is to provide consistent behavior when pytest extensions aren't present.

### CODE REVIEW

File: colcon_core/task/python/test/pytest.py [hunk 1]
```diff
@@ -71,23 +71,32 @@ def match(self, context, env, setup_py_data):  # noqa: D102
...
-            # https://github.com/pytest-dev/pytest-cov/blob/v2.5.0/CHANGELOG.rst
-            from pytest_cov import __version__ as pytest_cov_version
-            if parse_version(pytest_cov_version) >= parse_version('2.5.0'):
+            try:
+                from pytest_cov import __version__ as pytest_cov_version
+            except ImportError:
+                logger.warning(
+                    'Test coverage will not be produced for package '
+                    "'{context.pkg.name}' since the pytest extension 'cov' "
+                    'was not found'.format_map(locals()))
```
Comment (dirk-thomas) [inline colcon_core/task/python/test/pytest.py]: With / without the patch if the version of `cov` isn't new enough it is skipping the coverage report silently. So warning in the case where `cov` is not available seems inconsistent.

**Comment by cottsay** [comment]
I'd guess that I thought that an actionable message and actual test results were better than a backtrace if I was missing a system dependency.

The `setup.cfg` says:
```
  # the pytest dependency and its extensions are provided for convenience
  # even though they are only conditional
```
...but right now, if `pytest_repeat` or `pytest_rerunfailures` is missing, I get a message, but if `pytest_cov` is missing, I get a backtrace. But you're right, if the `package.xml` and the `setup.py`/`setup.cfg` agree that `python3-pytest-cov` is needed and rosdep was invoked to install it, it shouldn't happen.

I'm trying to figure out what this means for weak dependencies for `colcon_core`. As it is, the program performs correctly with any combination of pytest and its dependencies except if pytest is present but the cov plugin isn't, where the program crashes.

**Comment by dirk-thomas** [comment]
I am fine with adding the warning if `cov` isn't available at all (despite being declared as a test dependency in the package being tested) to make it consistent with the other optional dependencies. Though in the case that `cov` is available but not new enough it should also print a warning then.

**Comment by cottsay** [comment]
I added a warning, though it should be noted that only branch coverage information is withheld based on the `pytest_cov` version.

Thanks for taking a look.

File: colcon_core/task/python/test/pytest.py [hunk 2]
```diff
@@ -71,23 +71,38 @@ def match(self, context, env, setup_py_data):  # noqa: D102
...
+                        '--cov-branch',
+                    ]
+                else:
+                    logger.warning(
+                        'Test coverage will be produced but will not contain '
+                        'branch coverage information because the pytest '
+                        "extension 'cov' does not support it (need 2.5.0, "
+                        'have {pytest_cov_version})'.format_map(locals()))
+                env['COVERAGE_FILE'] = os.path.join(
+                    context.args.build_base, '.coverage')
```
Comment (dirk-thomas) [inline colcon_core/task/python/test/pytest.py]: Before the patch this environment variable was set even without branch coverage?

Comment (dirk-thomas) [reply colcon_core/task/python/test/pytest.py]: Nvm - still the same :wink:

**Pull request merged**